juju/juju
# Refactor charm state

Refactor the charm state to place the charm state as the core
concept and then move everything of that table into a charm_metadata
table. The charms "metadata" (metadata, actions, config, lxdprofile)
should be immutable. There are times where we want to register
a charm placeholder, but we don't have the charm. In these situations
we just want to add a row. This covers the case nicely. You either
have a charm row or not. That charm row can become available or not.

The only lingering thought is that the the charm reference_name, the
name part of the charm URL should also be on the charm table and
not on the charm_origin. I'm currently split on this, we can leave
it where it is and move it later. Most of the time it resides behind
a view anyway.

## Checklist





## QA steps

The tests should pass.

```sh
$ juju bootstrap lxd test
$ juju add-model default
$ juju deploy ubuntu
```

## Links

**Jira card:** [JUJU-6624](https://warthogs.atlassian.net/browse/JUJU-6624)
